diff -up sudo-1.8.6p3/plugins/sudoers/sssd.c.rulenames sudo-1.8.6p3/plugins/sudoers/sssd.c
--- sudo-1.8.6p3/plugins/sudoers/sssd.c.rulenames	2014-05-21 12:33:21.000768420 +0200
+++ sudo-1.8.6p3/plugins/sudoers/sssd.c	2014-05-21 12:38:13.779864718 +0200
@@ -1180,6 +1180,18 @@ sudo_sss_display_entry_long(struct sudo_
     int count = 0, i;
     debug_decl(sudo_sss_display_entry_long, SUDO_DEBUG_SSSD);
 
+    switch(handle->fn_get_values(rule, "cn", &val_array)) {
+    case 0:
+	if (val_array[0]) {
+	    lbuf_append(lbuf, _("\nSSSD Role: %s\n"), val_array[0]);
+	}
+	handle->fn_free_values(val_array);
+	val_array = NULL;
+	break;
+    default:
+	lbuf_append(lbuf, _("\nSSSD Role: UNKNOWN\n"));
+    }
+
     /* get the RunAsUser Values from the entry */
     lbuf_append(lbuf, "    RunAsUsers: ");
     switch (handle->fn_get_values(rule, "sudoRunAsUser", &val_array)) {
