diff -up ./plugins/sudoers/sudo_nss.c.display-privs ./plugins/sudoers/sudo_nss.c
--- ./plugins/sudoers/sudo_nss.c.display-privs	2017-01-13 23:30:15.000000000 -0500
+++ ./plugins/sudoers/sudo_nss.c	2017-08-31 07:41:02.764738698 -0400
@@ -348,7 +348,11 @@ display_privs(struct sudo_nss_list *snl,
     sudo_lbuf_destroy(&defs);
     sudo_lbuf_destroy(&privs);
 
-    debug_return_int(count > 0);
+/*
+ * This is ok, we return 1 which is success in this case
+ * and we don't want return failure even when there is nothing to print
+ */
+    debug_return_int(1);
 bad:
     sudo_lbuf_destroy(&defs);
     sudo_lbuf_destroy(&privs);
