diff -up sudo-1.8.6p3/src/tgetpass.c.sigpipefix sudo-1.8.6p3/src/tgetpass.c
--- sudo-1.8.6p3/src/tgetpass.c.sigpipefix	2015-03-03 10:23:23.219038693 +0100
+++ sudo-1.8.6p3/src/tgetpass.c	2015-03-03 10:23:43.089813184 +0100
@@ -173,7 +173,7 @@ restore:
     (void) sigaction(SIGTSTP, &savetstp, NULL);
     (void) sigaction(SIGTTIN, &savettin, NULL);
     (void) sigaction(SIGTTOU, &savettou, NULL);
-    (void) sigaction(SIGTTOU, &savepipe, NULL);
+    (void) sigaction(SIGPIPE, &savepipe, NULL);
     if (input != STDIN_FILENO)
 	(void) close(input);
 
