diff -up ./plugins/sudoers/regress/visudo/test2.err.ok.orig ./plugins/sudoers/regress/visudo/test2.err.ok
--- ./plugins/sudoers/regress/visudo/test2.err.ok.orig  2017-04-10 10:12:53.003000000 -0400
+++ ./plugins/sudoers/regress/visudo/test2.err.ok       2017-04-10 10:13:36.771000000 -0400
@@ -1 +1 @@
-visudo: stdin:1 cycle in User_Alias "FOO"
+Error: stdin:1 cycle in User_Alias "FOO"
diff -up ./plugins/sudoers/regress/visudo/test3.err.ok.orig ./plugins/sudoers/regress/visudo/test3.err.ok
--- ./plugins/sudoers/regress/visudo/test3.err.ok.orig  2017-04-10 10:13:12.141000000 -0400
+++ ./plugins/sudoers/regress/visudo/test3.err.ok       2017-04-10 10:13:56.842000000 -0400
@@ -1,2 +1,2 @@
-visudo: stdin:1 unused User_Alias "A"
-visudo: stdin:2 unused User_Alias "B"
+Warning: stdin:1 unused User_Alias "A"
+Warning: stdin:2 unused User_Alias "B"
